src/main/java/org/reactome/addlinks/dataretrieval/pharos/PharosDataRetriever.java

```diff
@@ -0,0 +1,195 @@
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.nio.file.Path;
+
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.StatusLine;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.ConnectTimeoutException;
+import org.apache.http.conn.HttpHostConnectException;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.reactome.release.common.dataretrieval.FileRetriever;
+
+/**
+ * Pharos data is accessed via a GraphQL API: https://pharos.nih.gov/api
+ * Data will be downloaded and then stored into simple text files.
+ *
+ * @author sshorser
+ *
+ */
+public abstract class PharosDataRetriever extends FileRetriever
+{
+	protected static final String SKIP_AMOUNT_TOKEN = "##SKIPAMOUNT##";
+
+	protected static final String TOP_AMOUNT_TOKEN = "##TOPAMOUNT##";
+
+	private static final String APPLICATION_JSON = "application/json";
+
+	// Pharos requested that we use the DEV endpoint for large requests, since it's not used by
+	// as many users, and I guess they don't want us slowing down the main endpoint.
+//	private static String pharosEndpoint = "https://ncatsidg-dev.appspot.com/graphql";
+//	private static String pharosEndpoint = "https://pharos-api.ncats.io/graphql"; // the non-dev endpoint.
+
+	// Some examples of testing Pharos' GraphQL endpoint.
+	// curl 'https://pharos-api.ncats.io/graphql' -H 'Accept-Encoding: gzip, deflate, br' -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Connection: keep-alive' -H 'DNT: 1' -H 'Origin: https://pharos-api.ncats.io' --data-binary '{"query":"query {\n  targets(filter: {facets: [{facet:\"Data Source\" values:[\"UniProt\"]}]}) {\n    count\n    filter\n    {\n      term\n    \tfacets {\n        facet\n        values\n      }\n    }\n    facets {\n      facet\n      values {\n        name\n        value\n      }\n    }\n    targets(top:100 skip:2 ) {\n      name\n      synonyms {\n        name\n        value\n      }\n    }\n  }\n}"}' --compressed
+	// curl 'https://ncatsidg-dev.appspot.com/graphql' -H 'Accept-Encoding: gzip, deflate, br' -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Connection: keep-alive' -H 'DNT: 1' -H 'Origin: https://ncatsidg-dev.appspot.com' --data-binary '{"query":"query {  targets {    targets(top:100) {      uniprot    }  }}\n"}' --compressed
+	private static int batchSize = 1000;
+
+	@Override
+	protected void doHttpDownload(Path path, HttpClientContext context) throws Exception, HttpHostConnectException, IOException
```

Comment:
Does it make sense to include `PharosDataException` in the method signature that is thrown from this method now?

Reply:
That was masked by the `throws Exception`. And yes, I would like to improve exception-handling in our applications in general, but no, this is not the time to do it.

Reply:
Yeah, but since HttpHostConnectException and IOException are included, it would make sense to include PharosDataException or all can be omitted if just Exception is kept.